test/integration/record_linkage.py

```diff
@@ -0,0 +1,211 @@
+    )
+
+    # Generate random number between 0-k many times and taking min
+    # This biases the number of edits to generally be low
+    num_edits = min(random.randrange(k) for i in range(10))
+    for _ in range(num_edits):
+        edit = edit_types[random.randrange(3)]
+        position = random.randrange(len(input_str) - 1)
+
+        if edit == "add":
+            input_list.insert(position, random.choice(characters))
+        elif edit == "delete":
+            input_list.pop(position)
+        else:
+            input_list[position] = random.choice(characters)
+
+    return "".join(input_list)
+
+
+def _noisify(col: pd.Series, sigma: float = 0.01, probability: float = 1) -> pd.Series:
+    """Add random noise to a column."""
+    noisy_rows = _RANDOM_GENERATOR.random(len(col)) > (1 - probability)
+    modifier_array = np.zeros(len(col))
+    modifier_array[noisy_rows] += _RANDOM_GENERATOR.normal(
+        scale=sigma, size=sum(noisy_rows)
+    )
+    return col + modifier_array
+
+
+def _modify_categorical(
+    df: pd.DataFrame, col: str, categories: list, probability: float = 0.01
+) -> pd.DataFrame:
+    """Randomly modify categorical column using given probability."""
+    modify_elements = _RANDOM_GENERATOR.random(len(df)) > (1 - probability)
+    df.loc[modify_elements, col] = _RANDOM_GENERATOR.choice(categories)
+    return df
+
+
+def _generate_random_test_df(
+    default_plant_name: str,
+    size: int = 2022 - 1994,
+    plant_name_max_edits: int = 5,
+    plant_type_error_prob: float = 0.01,
+    construction_type_error_prob: float = 0.01,
+    construction_year_error_prob: float = 0.01,
+    capacity_mean: float = 500.0,
+    capacity_sigma: float = 10.0,
+    capacity_change_prob: float = 0.01,
+    utility_id_error_prob: float = 0.01,
+    utility_id: int = random.randrange(1000),
+):
+    """Generate a random input DataFrame for testing record linkage."""
+
+    generated_df = pd.DataFrame(
+        {
+            "base_plant_name": [default_plant_name] * size,
+            "plant_type": [
+                random.choice(list(PLANT_TYPE_CATEGORIES["categories"].keys()))
+            ]
+            * size,
+            "report_year": list(range(1994, 1994 + size)),
+            "construction_type": [
+                random.choice(list(CONSTRUCTION_TYPE_CATEGORIES["categories"].keys()))
+            ]
+            * size,
+            "capacity_mw": [capacity_mean] * size,
+            "construction_year": [
+                random.randrange(
+                    VALID_PLANT_YEARS["lower_bound"], VALID_PLANT_YEARS["upper_bound"]
+                )
+            ]
+            * size,
+            "utility_id_ferc1": [utility_id] * size,
+        }
+    )
+
+    # Add random edits to plant name
+    generated_df["plant_name_ferc1"] = generated_df["base_plant_name"].apply(
+        _randomly_modify_string
+    )
+
+    # Modify capacity rows based on probability and sigma
+    generated_df["capacity_mw"] = _noisify(
```

Comment:
Probably not the best way to model capacity. Adding a bit of random noise to ~1% of rows doesn't account for expected capacity changes if generators were added or removed. It also doesn't really reflect cases where the value is just totally bogus